src/sessionManager.ts

```diff
@@ -0,0 +1,238 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+import { Binder, IBinderDelegate } from './binder';
+import DapConnection from './dap/connection';
+import { ITarget } from './targets/targets';
+import { IDisposable, IEvent, EventEmitter } from './common/events';
+import { DebugType } from './common/contributionUtils';
+import { TargetOrigin } from './targets/targetOrigin';
+import { TelemetryReporter } from './telemetry/telemetryReporter';
+import { ILogger } from './common/logging';
+import { Container } from 'inversify';
+import { createTopLevelSessionContainer } from './ioc';
+import { IChromeAttachConfiguration } from './configuration';
+
+/**
+ * Interface defining a strategy for obtaining a new DAP connection for
+ * a given session.
+ */
+export interface IConnectionStrategy {
+  init(telemetryReporter: TelemetryReporter, logger: ILogger): void;
```

Comment:
It looks like `init()` on the connection strategy can be reused between multiple/nested debug sessions. I think how it exists now might break nested sessions in VS Code. Maybe we should have an `IConnectionStrategyFactory` with an `.obtain()` call that returns something like

```ts
export interface IConnectionStrategy extends IDisposable /* to clean up any server */ {
  connection: Promise<DapConnection>;
}
```

Reply:
I don't think it would get reused since we're always creating a new instance in `createDebugAdapterDescriptor`, and the only place `init` is called is in the constructor of the session, which is only ever called from `createSession`. Am I missing something?

I do agree with making the connection strategy extend `IDisposable` though so we don't leak anything.

Reply:
I do think it's probably better to add another level of abstraction though to get away from the `init` method, which I'm not a fan of.

Reply:
@connor4312 I did a bit of refactoring, but still not fully happy with the connection strategy in general.

Is it ok if we merge this PR as is, and I'll have a follow-up coming shortly to focus on the connection piece?